README.md
<h1 align="center">unite</h1>

<p align="center">
  <a href="https://hex.pm/packages/unite">
    <img alt="hex.pm version" src="https://img.shields.io/hexpm/v/unite?style=flat-square"/>
  </a>
  <a href="LICENSE">
    <img alt="hex.pm license" src="https://img.shields.io/hexpm/l/unite?style=flat-square"/>
  </a>
  <img alt="erlang versions" src="https://img.shields.io/badge/erlang-21+-blue.svg?style=flat-square"/>
  <a href="https://github.com/sponsors/eproxus">
    <img alt="hex.pm license" src="https://img.shields.io/github/sponsors/eproxus?style=flat-square&color=%23ec6cb9"/>
  </a>
</p>

<p align="center">
  Pretty EUnit test formatters
</p>
<p align="center">
  <img src="assets/screenshots/screenshot.png" width="490" alt="Example output" />
</p>

## Installation & Usage

Add Unite as a dependency in your `rebar.config` and enable the Unite formatter:

```erlang
{profiles, [
    {test, [
        {deps, [unite]},
        {eunit_opts, [no_tty, {report, {unite_compact, []}}]}
    ]}
]}.

```

Using `no_tty` is important, because it disables the standard EUnit output.

Then just run Rebar 3 as usual: `rebar3 eunit`.

### Profiling

To get timings of long running tests, add `profile` to the option list:

```erlang
{eunit_opts, [no_tty, {report, {unite_compact, [profile]}}]}.
```

This will show the top 10 slowest tests (or the top N slowest tests if
`{profile, N}` is used).

## Donations

If you or your company use Unite and find it useful, [donations](https://github.com/sponsors/eproxus) are greatly appreciated!


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog], and this project adheres to
[Semantic Versioning].

# Change Log

## [Unreleased]

## [0.4.3] - 2023-09-27

### Fixed

- Fix crashing when formatting term containing multiple `~` control sequences
  (thanks @maxno-kivra!)

## [0.4.2] - 2023-09-05

### Fixed

- Handle bad test instantiators (thanks @rlipscombe & @danfilip!)

## [0.4.1] - 2023-09-04

### Fixed

- Fix crashing when formatting term diffs containing `~` control sequences

## [0.4.0] - 2022-11-29

### Added

- Print test case timeouts with message and stack trace

### Changed

- Drop support for Erlang/OTP 18, 19 and 20 (because of lack of error
  formatting APIs)

### Fixed

- Partial workaround for EUnit incorrectly reporting the wrong number of
  cancelled cases

## [0.3.2] - 2022-02-22

### Changed

- Drop support for Erlang R15, R16 and 17.0 (because of deprecated time APIs)

### Fixed

- Fixed header formatting inconsistencies
- Stack traces containing io:format control sequences now print correctly

## [0.3.1] - 2018-07-12

### Added

- Print something when no tests are run [\#13](https://github.com/eproxus/unite/issues/13)
- Enable backwards compatibility to R15 [\#12](https://github.com/eproxus/unite/issues/12)

### Fixed

- Crash on OTP 21 [\#19](https://github.com/eproxus/unite/issues/19)
- Crash in `format\_macro\_string` when attempting to pretty-print guard from `?assertMatch` [\#14](https://github.com/eproxus/unite/issues/14)
- Do not attempt to pretty-print expressions that cannot be parsed [\#15](https://github.com/eproxus/unite/pull/15) ([jonathanperret](https://github.com/jonathanperret))

## [0.3.0] - 2017-10-20

### Added

- Print text when no tests are run ([60cd086](https://github.com/eproxus/unite/commit/60cd086f2aac123988f5c1f5aa0a28df57e9c5fa))
- Support old time API for pre-18 versions ([998df2a](https://github.com/eproxus/unite/commit/998df2a2eb169fda13fae7f4739d46e2809ff7a2))

## [0.2.0] - 2017-09-12

### Added

- Show relative paths in stack traces [\#11](https://github.com/eproxus/unite/issues/11)
- Support OTP 18 [\#6](https://github.com/eproxus/unite/pull/6) ([knutin](https://github.com/knutin))

### Fixed

- Fix crash on unhandled case clause [\#10](https://github.com/eproxus/unite/pull/10) ([binarytemple](https://github.com/binarytemple))
- Fix spontaneous errors on strange testcase reports [\#5](https://github.com/eproxus/unite/pull/5) ([keynslug](https://github.com/keynslug))

## [v0.0.1] - 2015-08-25

### Fixed

- io:columns\(\) can return {error, enotsup} when run from an editor [\#1](https://github.com/eproxus/unite/issues/1)
- Fix for older Erlang version [\#2](https://github.com/eproxus/unite/pull/2) ([JonGretar](https://github.com/JonGretar))


[Unreleased]: https://github.com/eproxus/unite/compare/v0.4.3...HEAD
[0.4.3]: https://github.com/eproxus/unite/compare/v0.4.2...v0.4.3
[0.4.2]: https://github.com/eproxus/unite/compare/v0.4.1...v0.4.2
[0.4.1]: https://github.com/eproxus/unite/compare/v0.4.0...v0.4.1
[0.4.0]: https://github.com/eproxus/unite/compare/v0.3.2...v0.4.0
[0.3.2]: https://github.com/eproxus/unite/compare/v0.3.1...v0.3.2
[0.3.1]: https://github.com/eproxus/unite/compare/v0.3.0...v0.3.1
[0.3.0]: https://github.com/eproxus/unite/compare/v0.2.0...v0.3.0
[0.2.0]: https://github.com/eproxus/unite/compare/v0.0.1...v0.2.0
[0.0.1]: https://github.com/eproxus/unite/compare/048e7e0e84c34d06426a4d952973df7df307391a...v0.0.1

[Keep a Changelog]: https://keepachangelog.com/en/1.0.0/
[Semantic Versioning]: https://semver.org/spec/v2.0.0.html


test/unite_test.erl
-module(unite_test).

-include_lib("eunit/include/eunit.hrl").

%--- Demo Tests ----------------------------------------------------------------

-ifdef(DEMO).

assert_fail_test() ->
    io:format("THIS IS ONLY SHOWN IN FAILURES~n"),
    ?assert(false).

assert_test() ->
    io:format("SHOULD NOT BE VISIBLE~n"),
    ?assert(true).

assert_equal_test() ->
    ?assertEqual(
        #{
            value => 1,
            list => [a, b, c, d, e],
            tuple => {foo, bar},
            binary => <<"test string">>,
            map => #{a => 1, b => 2, c => value}
        },
        #{
            value => 2,
            list => [a, b, x, d, e],
            tuple => {foo, baz},
            binary => <<"test binary">>,
            map => #{foo => baz, bar => qux}
        }
    ).

assert_match_test() ->
    ?assertMatch(
       X when X < 1,
       2
    ).

error_test() -> error(error_in_test).

exit_test() -> exit(exit_in_test).

throw_test() -> throw(throw_in_test).

long_test() -> timer:sleep(250).

short_test() -> timer:sleep(50).

% EUnit has a bug where not all of the below cases are catched. To see each
% individual case for sure, comment out the others:

setup_test_() ->
    {setup, fun() -> error(error_in_setup) end, fun(_) -> ok end, [?_assert(true)]}.

cleanup_test_() ->
    {setup, fun() -> ok end, fun(_) -> error(error_in_cleanup) end, [?_assert(true)]}.

instantiation_test_() ->
    {setup, fun() -> ok end, fun(_) -> ok end, fun(_) -> error(error_in_instantiation) end}.

bad_spec_test_() -> {foo, bar, baz}.

single_module_not_found_test_() -> foo.

module_not_found_test_() -> {foo, bar}.

function_not_found_test_() -> {?MODULE, bar}.

generator_test_() ->
    error(error_in_generator).

-endif.


test/unite_demo_test.erl
% NOTE: This is a demo test module, only for output purposes. Use like this:
%
%    $ rebar3 as demo eunit

-module(unite_demo_test).

-include_lib("eunit/include/eunit.hrl").

%--- Demo Tests ----------------------------------------------------------------

-ifdef(DEMO).

assert_fail_test() ->
    io:format("THIS IS ONLY SHOWN IN FAILURES~n"),
    ?assert(false).

assert_test() ->
    io:format("SHOULD NOT BE VISIBLE~n"),
    ?assert(true).

assert_equal_test() ->
    ?assertEqual(
        #{
            value => 1,
            list => [a, b, c, d, e],
            tuple => {foo, bar},
            binary => <<"test string ~p ~s">>,
            map => #{a => 1, b => 2, c => value}
        },
        (fun() -> #{
            value => 2,
            list => [a, b, x, d, e],
            tuple => {foo, baz},
            binary => <<"test binary ~p ~s">>,
            map => #{foo => baz, bar => qux}
        } end)()
    ).

assert_match_test() ->
    ?assertMatch(
       X when X < 1,
       (fun() -> 2 end)()
    ).

assert_throw_test() ->
    ?assertThrow(foo, throw(foo)),
    ?assertThrow(foo, throw(bar)).

assert_error_test() ->
    ?assertError(foo, error(foo)),
    ?assertError(foo, error(bar)).

assert_exit_test() ->
    ?assertExit(foo, exit(foo)),
    ?assertExit(foo, exit(bar)).

error_test() -> error(error_in_test).

exit_test() -> exit(exit_in_test).

throw_test() -> throw(throw_in_test).

long_test() -> timer:sleep(250).

short_test() -> timer:sleep(50).

bad_test_instantiator_test_() -> {foreach, fun() -> ok end, [fun(_) -> ok end]}.

io_format_test() ->
    % If a control code sneaks into the stack trace, we have to make sure we can
    % io:format it properly when Unite pretty prints:
    apply(io, format, ["~p ~s", []]).

% Generators

setup_test_() ->
    {setup, fun() -> error(error_in_setup) end, fun(_) -> ok end, [?_assert(true)]}.

cleanup_test_() ->
    {setup, fun() -> ok end, fun(_) -> error(error_in_cleanup) end, [?_assert(true)]}.

instantiation_test_() ->
    {setup, fun() -> ok end, fun(_) -> ok end, fun(_) -> error(error_in_instantiation) end}.

timeout_test_() ->
    {timeout, 0, [fun() -> timer:sleep(1000) end]}.

% EUnit has a bug where the below cases are not included if the above
% generators fail. To see these, comment out the generators above and every
% other test case below:

bad_spec_test_() -> {foo, bar, baz}.

single_module_not_found_test_() -> foo.

module_not_found_test_() -> {foo, bar}.

function_not_found_test_() -> {?MODULE, bar}.

generator_test_() ->
    error(error_in_generator).

-endif.


src/unite.erl
-module(unite).

% API
-export([test/1]).

%--- API ----------------------------------------------------------------------

test(Specs) -> test(Specs, []).

test(Specs, Options) ->
    Format = proplists:get_value(format, Options, unite_verbose),
    Tests = discover_tests(Specs),
    {Format, {Passed, Failed}} = run_tests(Tests, {Format, {0, 0}}),
    Format:test_summary(Passed, Failed, 0).

%--- Internal Functions -------------------------------------------------------

discover_tests(Specs) when is_list(Specs) ->
    [discover_tests(S) || S <- Specs];
discover_tests(Module) when is_atom(Module) ->
    [discover_tests({Module, F, A}) || {F, A} <- Module:module_info(exports)];
discover_tests({M, F, A}) ->
    case ends_with(F, "_test") of
        true  -> fun M:F/A;
        false ->
            case ends_with(F, "_test_") of
                true  -> discover_tests(M:F());
                false -> []
            end
    end;
discover_tests({Line, Fun}) when is_integer(Line), is_function(Fun) ->
    Fun.

ends_with(Atom, String) ->
    Binaries = [atom_to_binary(Atom, latin1), list_to_binary(String)],
    case binary:longest_common_suffix(Binaries) of
        L when L == length(String) -> true;
        _                          -> false
    end.

run_tests([{M, F, 0}|Tests], State) ->
    run_tests(Tests, run_test(fun M:F/0, State));
run_tests([Deep|Tests], State) when is_list(Deep) ->
    run_tests(Tests, run_tests(Deep, State));
run_tests([Fun|Tests], State) when is_function(Fun) ->
    run_tests(Tests, run_test(Fun, State));
run_tests([], State) ->
    State.

run_test(F, {Format, {Passed, Failed}}) ->
    Result = try
        ok = F(),
        {passed, ok}
    catch
        T:R ->
            {failed, {T, R, erlang:get_stacktrace()}}
    end,

    Info = erlang:fun_info(F),
    Format:test_case(
        Result,
        % element(random:uniform(2), {Result, {skipped, {a, b, []}}}),
        proplists:get_value(module, Info),
        proplists:get_value(name, Info)
    ),
    {Format, {Passed + 1, Failed}}.


src/unite_verbose.erl
-module(unite_verbose).

% API
-export([test_summary/3]).
-export([test_case/3]).

% Clear line: "\e[2K"

%--- API ----------------------------------------------------------------------

test_summary(Passed, Failed, Skipped) ->
    io:format("~n~s~n", [iolist_to_binary(iojoin([
        non_zero(Passed, green, [i2b(Passed), " tests passed"]),
        non_zero(Failed, red, [i2b(Failed), " tests failed"]),
        non_zero(Skipped, yellow, [i2b(Skipped), " tests skipped"])
    ], "  "))]).

test_case({Result, _}, Module, Function) ->
    io:format("~s ~p:~p~n", [result(Result), Module, Function]).

%--- Internal Functions -------------------------------------------------------

i2b(Integer) -> integer_to_binary(Integer).

non_zero(Int, Colors, IOData) ->
    case Int of
        0 -> [];
        _ -> colorize(IOData, Colors)
    end.

colorize(String, []) ->
    String;
colorize(String, [Color|Colors]) ->
    colorize(color:Color(String), Colors);
colorize(String, Color) when is_atom(Color) ->
    colorize(String, [Color]).

iojoin([], _Separator)         -> [];
iojoin([[]|List], Separator)   -> iojoin(List, Separator);
iojoin([Item], _Separator)     -> Item;
iojoin([Item|List], Separator) -> [Item, Separator, iojoin(List, Separator)].

result(passed) -> color:green(string:right("Passed", 7));
result(failed) -> color:red(string:right("Failed", 7));
result(skipped) -> color:yellow(string:right("Skipped", 7)).


src/unite_compact.erl
% @private
-module(unite_compact).

-behaviour(eunit_listener).

% EUnit Callbacks
-export([start/0]).
-export([start/1]).
-export([init/1]).
-export([handle_begin/3]).
-export([handle_end/3]).
-export([handle_cancel/3]).
-export([terminate/2]).

-export([ioindent/2]).

% Clear line: "\e[2K"

-record(s, {
    start = current_time(),
    cases = [],
    profile = false,
    profile_max = 10
}).

%--- Macros --------------------------------------------------------------------

-ifdef(OTP_RELEASE).
-define(ERROR_FORMATTER, erl_error).
-else.
-define(ERROR_FORMATTER, lib).
-endif.

%--- EUnit Callbacks -----------------------------------------------------------

start() ->
    start([]).

start(Options) ->
    eunit_listener:start(?MODULE, Options).

init(Options) ->
    case get(profile, Options) of
        undefined ->
            #s{};
        true ->
            #s{profile = true};
        Max when is_integer(Max), Max >= 0 ->
            #s{profile = true, profile_max = Max}
    end.

handle_begin(_Type, _Data, State) ->
    State.

handle_end(test, Data, State) ->
    case get(status, Data) of
        ok            -> format(color:green("."));
        skip          -> format(color:yellow("S"));
        {skipped, _ } -> format(color:yellow("S"));
        {error, _}    -> format(color:redb("F"))
    end,
    State#s{cases = State#s.cases ++ [Data]};
handle_end(group, _Data, State) ->
    State.

handle_cancel(group, Data, State) ->
    case get(reason, Data) of
        undefined ->
            State;
        _Else ->
            State#s{cases = State#s.cases ++ [Data]}
    end;
handle_cancel(test, Data, State) ->
    format(color:yellow("C")),
    NewData = case get(status, Data) of
        undefined -> [{status, {cancelled, undefined}}|Data];
        _Else     -> Data
    end,
    State#s{cases = State#s.cases ++ [NewData]}.

terminate({ok, Result}, #s{cases = Cases} = State) ->
    Grouped = group_by(
        fun(C) ->
            case get(status, C) of
                {error, _} ->
                    failed;
                {skipped, _} ->
                    failed;
                {cancelled, _} ->
                    failed;
                ok ->
                    succesful;
                _Status  ->
                    case get(reason, C) of
                        {abort, _} -> failed;
                        {blame, _Case} -> blamed
                    end
            end
        end,
        Cases
    ),
    print_failures(Grouped),
    print_times(State),
    print_summary(Result, State, Grouped).

%--- Internal Functions -------------------------------------------------------

print_failures(#{failed := Failures}) ->
    Indexed = lists:zip(lists:seq(1, length(Failures)), Failures),
    [print_failure(I, F) || {I, F} <- Indexed],
    format("~n");
print_failures(_Cases) ->
    ok.

% Individual Test Case

print_failure(Index, Failure) ->
    Reason = get(reason, Failure),
    Info = get(status, Failure, Reason),

    {Header, Details} = format_info(Failure, Info),
    format("~n~n ~p) ~s~n", [Index, Header]),
    format(ioindent(4, Details)),
    case format_output(Failure) of
        undefined -> ok;
        Output    -> format("~n~s", [ioindent(4, Output)])
    end.

format_info(Failure, {error, {error, {assert, Info}, ST}}) ->
    Expr = get(expression, Info),
    {
        color:red(format_case(Failure, ST)),
        [color:redb("Assert failed: "), format_macro_string(Expr)]
    };
format_info(Failure, {error, {error, {assertEqual, Info}, ST}}) ->
    Expected = get(expected, Info),
    Actual = get(value, Info),
    Exp = diff_prep_term(Expected),
    Act = diff_prep_term(Actual),
    Diff = tdiff:diff(Exp, Act),
    {
        color:red(format_case(Failure, ST)),
        io_lib:format("~s ~s~n~s", [
            color:redb("Assert equal failed:"),
            [
                color:blueb("-Expected-"),
                " ",
                color:yellowb("+Actual+")
            ],
            format_diff(Diff)
        ])
    };
format_info(Failure, {error, {error, {assertMatch, Info}, ST}}) ->
    Expr = get(expression, Info),
    Pattern = get(pattern, Info),
    Value = get(value, Info),
    {
        color:red(format_case(Failure, ST)),
        io_lib:format("~s~n~s~n~s~n~s~n~s~n~s~n~s~n", [
            color:redb("Assert match failed:"),
            color:magentab("Expression:"),
            ioindent(4, format_macro_string(Expr)),
            color:blueb("Pattern:"),
            ioindent(4, format_macro_string(Pattern)),
            color:yellowb("Actual:"),
            ioindent(4, format_term(Value, 0, 8))
        ])
    };
format_info(Failure, {error, {error, {assertException, Info}, ST}}) ->
    case get(unexpected_exception, Info) of
        undefined ->
            Success = get(unexpected_success, Info),
            Term = format_term(Success, 22, 4),
            {
                color:red(format_case(Failure, ST)),
                case multiline(Term) of
                    true ->
                        io_lib:format("~s~n~s", [
                            color:redb("Unexpected success:"),
                            color:red(format_term(Success, 0, 4))
                        ]);
                    false ->
                        [
                            color:redb("Unexpected success:"),
                            " ",
                            color:red(Term)
                        ]
                end
            };
        {E, R, NewST} ->
            {
                color:red(format_case(Failure, ST)),
                io_lib:format("~s~n~s", [
                    color:redb("Unexpected exception:"),
                    color:red(format_exception(E, R, NewST))
                ])
            }
    end;
format_info(Failure, {error, {E, R, ST}}) ->
    {
        color:red(format_case(Failure, ST, red)),
        [
            color:redb("Uncaught exception:"),
            io_lib:format("~n", []),
            color:red(format_exception(E, R, ST))
        ]
    };
format_info(_Failure, {abort, {bad_test, Test}}) ->
    {
        color:yellow("Bad test specification:"),
        [
            color:yellowb("Invalid specification: "),
            color:yellow(io_lib:format("~p", [Test]))
        ]
    };
format_info(Failure, {abort, {generator_failed, {MFA, {E, R, ST}}}}) ->
    {
        color:yellow(format_case(Failure, [add_info(MFA, ST)])),
        [
            color:yellowb("Generator failed:"),
            io_lib:format("~n", []),
            color:yellow(format_exception(E, R, ST))
        ]
    };
format_info(Failure, {abort, {{bad_instantiator, {{M, F, A}, Result}}, {E, R, ST}}}) ->
    Location = {M, F, A, []},
    {
        color:yellow(format_case(Failure, [Location|ST])),
        [
            color:yellowb(io_lib:format("Bad test instantiator: ~p~n", [
                Result
            ])),
            color:yellow(format_exception(E, R, ST))
        ]
    };
format_info(Failure, {abort, {Reason, {E, R, ST}}}) ->
    {
        color:yellow(format_case(Failure, ST)),
        [
            color:yellowb(case Reason of
                setup_failed -> "Setup failed: ";
                cleanup_failed -> "Cleanup failed: ";
                instantiation_failed -> "Instantiation failed: ";
                Other -> io_lib:format("Unknown failure reason (~p): ", [Other])
            end),
            io_lib:format("~n", []),
            color:yellow(format_exception(E, R, ST))
        ]
    };
format_info(_Failure, {skipped, Reason}) ->
    {
        color:yellow("Bad test specification:"),
        [
            color:yellowb("Specification exception: "),
            color:yellow(io_lib:format("~p", [Reason]))
        ]
    };
format_info(_Failure, {abort, {module_not_found, Module}}) ->
    {
        color:yellow("Bad test specification:"),
        [
            color:yellowb("Specification exception: "),
            color:yellow(io_lib:format("~p", [{module_not_found, Module}]))
        ]
    };
format_info(Failure, {cancelled, undefined}) ->
    case lists:keyfind(reason, 1, Failure) of
        {reason, {timeout, #{stacktrace := ST}}} ->
            {
                color:yellow(format_case(Failure, [])),
                [
                    color:yellowb("Test case timeout!"),
                    io_lib:format("~n", []),
                    color:yellow(format_exception(exit, killed, ST))
                ]
            };
        Reason ->
            {
                color:yellow(format_case(Failure, [])),
                [
                    color:yellowb("Unknown EUnit error!"),
                    io_lib:format("~n", []),
                    color:black(io_lib:format("~p", [Reason]))
                ]
            }
    end.

diff_prep_term(Term) ->
    Pretty = format_term(Term, 0, 0),
    TermSplit = "([,\\[\\]\\{\\}]|\\s+=>\\s+|#\\{)",
    re:split(Pretty, TermSplit, [trim]).

format_term(Term, Indent, Outer) ->
    io_lib_pretty:print(Term, Indent, columns() - Outer, -1).

format_diff([]) ->
    [];
format_diff([{eq, Str}|Rest]) ->
    [Str|format_diff(Rest)];
format_diff([{del, Str}|Rest]) ->
    [color:blue(["-", Str, "-"])|format_diff(Rest)];
format_diff([{ins, Str}|Rest]) ->
    [color:yellow(["+", Str, "+"])|format_diff(Rest)].


format_case(Failure, ST) -> format_case(Failure, ST, white).

format_case(Failure, ST, Color) ->
    case get(desc, Failure) of
        undefined -> format_source(Failure, ST);
        Desc ->
            io_lib:format("~s~n~s", [
                colorize([$", Desc, $"], cyan),
                colorize(ioindent(4, format_source(Failure, ST)), Color)
            ])
    end.

format_source(Failure, ST) ->
    case get(source, Failure) of
        undefined ->
            format_stack_line(ST);
        MFA ->
            format_stack_line([add_info(MFA, ST)])
    end.

format_stack_line([{M, F, A, I} | _]) ->
    case {get(file, I), get(line, I)} of
        {undefined, undefined} ->
            io_lib:format("~p:~p/~p", [M, F, A]);
        {File, L} ->
            io_lib:format("~p/~p (~s:~p)", [F, A, relative_file(File), L])
    end;
format_stack_line([]) ->
    "unknown location".

format_exception(Error, Reason, Stacktrace) ->
    format_exception(1, Error, Reason, relative_stacktrace(Stacktrace),
        fun(_M, _F, _A) -> false end,
        fun(T, I) ->
            io_lib_pretty:print(T, I, columns(), -1)
        end
    ).

format_output(Failure) ->
    case get(output, Failure) of
        <<>> -> undefined;
        [<<>>] -> undefined;
        undefined -> undefined;
        Output ->
            [
                color:blackb("Output:"),
                io_lib:format("~n", []),
                ioindent(2, Output)
            ]
    end.

format_macro_string(Str) ->
    case lists:member($?, Str) of
        true ->
            [C || C <- Str, C =/= $ ];
        false ->
            {ok, S, _} = erl_scan:string(Str ++ "."),
            case erl_parse:parse_exprs(S) of
              {ok, P} -> erl_pp:exprs(P);
              _ -> Str
            end
    end.

% Profiling

print_times(#s{profile_max = Max, cases = Cases, profile = P}) when P ->
    Times = [{T, format_case(C, [])} || C <- Cases, T <- [get(time, C)], is_integer(T)],
    Top = lists:sublist(lists:reverse(lists:sort(Times)), Max),
    case length(Top) of
        0 ->
            ok;
        N ->
            Title = colorize(io_lib:format("Top ~p slowest tests:", [N]), yellow),
            format("~n~n  ~s~n", [Title]),
            [print_time(T, C) || {T, C} <- Top]
    end;
print_times(_State) ->
    ok.

print_time(Ms, Case) ->
    Time = colorize(string:left(format_time(Ms), 10), red),
    format("    ~s~n      ~s~n", [Case, Time]).

% Summary

print_summary(Result, State, Grouped) ->
    case get_all(Result, [pass, fail, skip, cancel]) of
        [0, 0, 0, 0] ->
            format("0 tests run~n");
        [Pass, Fail, Skip, TotalCancel] ->
            Cancel = TotalCancel - length(maps:get(blamed, Grouped, [])),
            Ms = elapsed_millisecond(State#s.start),
            Time = format_time(Ms),
            format("~n~s~n", [iolist_to_binary(iojoin([
                non_zero(Pass, green, plural(Pass, "test", "passed")),
                non_zero(Fail, red, plural(Fail, "test", "failed")),
                non_zero(Skip, yellow, plural(Skip, "test", "skipped")),
                non_zero(Cancel, yellow, plural(Cancel, "fixture", "cancelled")),
                color:blackb(io_lib:format("(~s)", [Time]))
            ], "  "))])
    end.

plural(Number, Noun, Postfix) ->
    Text = case Number of 1 -> Noun; Number -> [Noun, "s"] end,
    [i2b(Number), " ", Text, " ", Postfix].

% Utilities

format_time(Ms) -> io_lib:format("~.2f s", [Ms / 1000]).

get_all(Proplist, Keys) ->
    [get(K, Proplist) || K <- Keys].

i2b(Integer) -> list_to_binary(integer_to_list(Integer)).

non_zero(Int, Colors, IOData) ->
    case Int of
        0 -> [];
        _ -> colorize(IOData, Colors)
    end.

colorize(String, []) ->
    String;
colorize(String, [Color|Colors]) ->
    colorize(color:Color(String), Colors);
colorize(String, Color) when is_atom(Color) ->
    colorize(String, [Color]).

iojoin([], _Separator)         -> [];
iojoin([[]|List], Separator)   -> iojoin(List, Separator);
iojoin([Item], _Separator)     -> Item;
iojoin([Item|List], Separator) -> [Item, Separator, iojoin(List, Separator)].

ioindent(Indent, IOData) when is_integer(Indent) ->
    Spacing = iolist_to_binary(lists:duplicate(Indent, 32)),
    [Spacing, ioindent(Spacing, IOData)];
ioindent(Spacing, [10|IOData]) ->
    [10, Spacing|ioindent(Spacing, IOData)];
ioindent(Spacing, Binary) when is_binary(Binary) ->
    binary:replace(Binary, <<"\n">>, <<"\n", Spacing/binary>>, [global]);
ioindent(Spacing, [Sub|IOData]) ->
    [ioindent(Spacing, Sub)|ioindent(Spacing, IOData)];
ioindent(_Spacing, []) ->
    [];
ioindent(_Spacing, Other) ->
    Other.

add_info({M, F, A}, [])                 -> {M, F, A, []};
add_info({M, F, A}, [{M, _, _, I}|_ST]) -> {M, F, A, I};
add_info(MFA, [_Line|ST])               -> add_info(MFA, ST).

multiline([10|_IOData]) ->
    true;
multiline([List|IOData]) when is_list(List) ->
    multiline(List) orelse multiline(IOData);
multiline([_|IOData]) ->
    multiline(IOData);
multiline(IOData) when is_binary(IOData) ->
    binary:match(IOData, <<"\n">>) =/= nomatch;
multiline([]) ->
    false.

get(Key, Proplist)          -> proplists:get_value(Key, Proplist).
get(Key, Proplist, Default) -> proplists:get_value(Key, Proplist, Default).

columns() -> case io:columns() of {ok, Columns} -> Columns; _Error -> 80 end.

relative_stacktrace([{M, F, A, Opts}|Stacktrace]) ->
    [{M, F, A, relative_stack_opts(Opts)}|relative_stacktrace(Stacktrace)];
relative_stacktrace([]) ->
    [].

relative_stack_opts([{file, File}|Opts]) -> [{file, relative_file(File)}|Opts];
relative_stack_opts([Opt|Opts])          -> [Opt|Opts];
relative_stack_opts([])                  -> [].

relative_file(File) ->
    {ok, CWD} = file:get_cwd(),
    case {filename:split(CWD), filename:split(File)} of
        {[Root|_] = A, [Root|_] = B} -> filename:join(relative_file(A, B));
        {_, _}                       -> File
    end.

relative_file([P|A], [P|B]) -> relative_file(A, B);
relative_file([_|A], B)     -> [".."|relative_file(A, B)];
relative_file([], B)        -> B.

current_time() -> erlang:monotonic_time().

elapsed_millisecond(Start) ->
    Elapsed = current_time() - Start,
    erlang:convert_time_unit(Elapsed, native, millisecond).

format(Format)       -> format("~ts", [Format]).
format(Format, Data) -> io:format(iolist_to_binary(Format), Data).

format_exception(I, Class, Reason, StackTrace, StackFun, FormatFun) ->
    ?ERROR_FORMATTER:format_exception(I, Class, Reason, StackTrace, StackFun, FormatFun).

group_by(Fun, List) ->
    lists:foldl(fun(Item, Acc) ->
        maps:update_with(Fun(Item), fun(Items) -> Items ++ [Item] end, [Item], Acc)
    end, #{}, List).
